projenrc/gen-api-resource.ts

```diff
@@ -0,0 +1,284 @@
+import * as fs from 'fs';
+import * as path from 'path';
+import { Project, SourceCode } from 'projen';
+import { snakeCase } from 'snake-case';
+
+/**
+ * Generate a source file for the `ApiResource` module, based on information about API
+ * resources generated by running `kubectl api-resources -o wide`.
+ *
+ * We do this because `kubectl api-resources` doesn't support JSON output
+ * formatting, and at the time of writing, parsing this command output seemed
+ * simpler than extracting information from the OpenAPI schema.
+ */
+export function generateApiResources(project: Project, sourcePath: string, outputPath: string) {
+  const resourceTypes = parseApiResources(sourcePath);
+  const ts = new SourceCode(project, outputPath);
+  if (ts.marker) {
+    ts.line(`// ${ts.marker}`);
+  }
+  ts.line();
+  ts.line('/**');
+  ts.line(' * Represents a resource or collection of resources.');
+  ts.line(' */')
+  ts.open('export interface IApiResource {');
+  ts.line('/**');
+  ts.line(' * The group portion of the API version (e.g. `authorization.k8s.io`).');
+  ts.line(' */')
+  ts.line('readonly apiGroup: string;')
+  ts.line();
+  ts.line('/**');
+  ts.line(' * The name of a resource type as it appears in the relevant API endpoint.');
+  ts.line(' * @example - "pods" or "pods/log"');
+  ts.line(' * @see https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-resources');
+  ts.line(' */');
+  ts.line('readonly resourceType: string;');
+  ts.line();
+  ts.line('/**');
+  ts.line(' * The unique, namespace-global, name of an object inside the Kubernetes cluster');
+  ts.line(' */')
+  ts.line('readonly name?: string;')
```

Comment:
For consistency:

```suggestion
  ts.line('readonly resourceName?: string;')
```

Also maybe add something like:

*if this is omitted, the ApiResource should represent all objects of the given type.*